closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java

```diff
@@ -416,16 +467,23 @@ int decryptFinal(byte[] in, int inOfs, int len,
                      byte[] out, int outOfs)
         throws IllegalBlockSizeException, AEADBadTagException,
         ShortBufferException {
+        try {
+            len = Math.addExact(len,ibuffer.size());
+        }
+        catch(ArithmeticException e) {
+            throw new ProviderException("SunJCE provider only supports " +
+                "input size up to " + MAX_BUF_SIZE + " bytes");
+        }
```

Comment:
I would prefer to see explicit range checking:
```
  (0 <= len) && (len <= MAX_BUF_SIZE - ibuffer.size())
```
As written, there's no clear connection between the situations in which `addExact` would throw an exception and the value of `MAX_BUF_SIZE`.

Reply:
The code I add here is not intended to be connected to `MAX_BUF_SIZE`. I need to execute: `len += ibuffer.size()`. However, since this might overflow, I add the code above to handle the possibility of overflow.

The range checking you mention int he above comment is found in the function `checkDataLength`, which is called on line 484. The only part of the range checking you mention above that is not currently there is `len >= 0`. Hence, I added this check at the beginning of `decryptFinal`:

```java
if (len < 0) {
  throw new ProviderException("Input len is negative"); }
```

Reply:
I suggest `len` should not be modified at all in this method. Instead, after the check you added that `len` is not negative, the expression `ibuffer.size() < tagLenBytes - len` is safe from overflow and does the necessary validation.

Parameters `inOfs` and `outOfs` should also be validated.

Reply:
Sure, I can remove it. I was just thinking because in the original code we also have the following:

```java
if (out.length - outOfs < ((ibuffer.size() + len) - tagLenBytes)) {
  throw new ShortBufferException("Output buffer too small");
}
```

This was already a potential integer overflow. But I can change that condition as well to: `out.length - outOfs - ibuffer.size() < len - tagLenBytes`

Reply:
This need not change: the test above that `ibuffer.size() < tagLenBytes - len` ensures that `ibuffer.size() + len` cannot overflow.

Reply:
Makes sense. So I changed the code to have the following at the beginning:
```java
if (len < 0 && (len > MAX_BUF_SIZE - ibuffer.size() + tagLen)) {
  throw new ProviderException("SunJCE provider only supports " +
  "a positive input size up to " + MAX_BUF_SIZE + " bytes");    
}
```
This enforces the condition you suggested initially in this thread. I also removed the `checkDataLength` call since its redundant now (with the extra check above).